try-slang.js

```diff
@@ -71,21 +82,74 @@ async function webgpuInit()
         return;
     }
     context = configContext(device, canvas);
+
+    // The default resolution of a canvas element is 300x150, which is too small compared to the container size of the canvas,
+    // therefore, we have to set the resolution same as the container size.
+
+    const observer = new ResizeObserver((entries) => { resizeCanvasHandler(entries); });
+    observer.observe(canvas);
+}
+
+function resizeCanvas(entries)
+{
+    const canvas = entries[0].target;
+
+    const width = canvas.clientWidth;
+    const height = canvas.clientHeight;
+
+    if (canvas.width !== width || canvas.height !== height)
+    {
+        // ensure the size won't be 0 nor exceed the limit, otherwise WebGPU will throw an errors
+        canvas.width = Math.max(1, Math.min(width, device.limits.maxTextureDimension2D));
+        canvas.height = Math.max(1, Math.min(height, device.limits.maxTextureDimension2D));
+
+        return true;
+    }
+
+    return false;
+}
+
+// We use the timer in the resize handler debounce the resize event, otherwise we could end of rendering
+// multiple useless frames.
+function resizeCanvasHandler(entries)
+{
+    setTimeout(() => {
+        var needResize = resizeCanvas(entries);
+        if (needResize)
+        {
+            if (computePipeline && passThroughPipeline)
+            {
+                computePipeline.createOutput(true);
+                computePipeline.createBindGroup();
+                passThroughPipeline.inputTexture = computePipeline.outputTexture;
+                passThroughPipeline.createBindGroup();
+                requestAnimationFrame(render);
+            }
+        }
+    }, 100);
 }
 
 
-function render(shaderCode)
+async function render(timeMS)
 {
-    const module = device.createShaderModule({code:shaderCode});
-    computePipeline.createPipeline(module);
+    // we only need to re-create the pipeline when the source code is changed and recompiled.
+    if (sourceCodeChange)
+    {
+        var shaderCode = codeGenArea.getValue();
+        const module = device.createShaderModule({code:shaderCode});
+        computePipeline.createPipeline(module);
+        sourceCodeChange = false;
+    }
 
+
+    computePipeline.updateUniformBuffer(timeMS * 0.01);
     // Encode commands to do the computation
     const encoder = device.createCommandEncoder({ label: 'compute builtin encoder' });
     const pass = encoder.beginComputePass({ label: 'compute builtin pass' });
 
     pass.setBindGroup(0, computePipeline.bindGroup);
     pass.setPipeline(computePipeline.pipeline);
-    pass.dispatchWorkgroups(canvas.width, canvas.height);
+    pass.dispatchWorkgroups(canvas.clientWidth, canvas.clientHeight);
```

Comment:
Todo: Need reflection and diversity client size by thread group's size.

Reply:
Right.
There is another location that needs the reflection API which is the binding point. Now I can only access the generated code to find out the binding point.